device-connectors/src/testflinger_device_connectors/devices/__init__.py

```diff
@@ -222,6 +222,44 @@ def allocate(self, args):
         with open("device-info.json", "w", encoding="utf-8") as devinfo_file:
             devinfo_file.write(json.dumps(device_info))
 
+    def import_ssh_key(self, key: str) -> None:
+        """Import SSH key provided in Reserve data.
+
+        :param key: SSH key to import.
+        :raises RuntimeError: If failure during import ssh keys
+        """
+        cmd = ["ssh-import-id", "-o", "key.pub", key]
+        for retry in range(10):
+            try:
+                subprocess.run(
+                    cmd,
+                    timeout=30,
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.STDOUT,
+                    check=True,
+                )
+                logger.info("Successfully imported key: %s", key)
+                break
+            except (
+                subprocess.CalledProcessError,
+                subprocess.TimeoutExpired,
+            ) as exc:
+                if isinstance(exc, subprocess.CalledProcessError):
+                    output = (exc.stdout or b"").decode()
+                    if "status_code=404" in output:
+                        raise RuntimeError(
+                            f"Failed to import ssh key: {key}. User not found."
+                        ) from exc
+
+                # If any other error, attempt to retry
+                logger.error("Unable to import ssh key from: %s", key)
+                logger.info("Retrying...")
+                time.sleep(min(2**retry, 30))
```

Comment:
Since the `try` block `break`s out of the loop anyways, let's move the logging and sleep blocks out of the `except` block and split the `except` block into two blocks so we don't have to check `isinstance`.

```suggestion

            except subprocess.TimeoutExpired:
                pass
            except subprocess.CalledProcessError as exc:
                output = (exc.stdout or b"").decode()
                if "status_code=404" in output:
                    raise RuntimeError(
                        f"Failed to import ssh key: {key}. User not found."
                    ) from exc

            logger.error("Unable to import ssh key from: %s", key)
            logger.info("Retrying...")
            time.sleep(min(2**retry, 30))
```